blockchain/blockindex.go

```diff
@@ -0,0 +1,501 @@
+func (bi *blockIndex) HaveBlock(hash *chainhash.Hash) bool {
+	bi.RLock()
+	_, hasBlock := bi.index[*hash]
+	bi.RUnlock()
+	return hasBlock
+}
+
+// loadBlockNode loads the block identified by hash from the block database,
+// creates a block node from it, and updates the memory block chain accordingly.
+// It is used mainly to dynamically load previous blocks from the database as
+// they are needed to avoid needing to put the entire block chain in memory.
+//
+// This function MUST be called with the block index lock held (for writes).
+// The database transaction may be read-only.
+func (bi *blockIndex) loadBlockNode(dbTx database.Tx, hash *chainhash.Hash) (*blockNode, error) {
+	// Load the block from the db.
+	block, err := dbFetchBlockByHash(dbTx, hash)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the new block node for the block.
+	blockHeader := block.MsgBlock().Header
+	node := newBlockNode(&blockHeader, stake.FindSpentTicketsInBlock(block.MsgBlock()))
+	node.inMainChain = true
+
+	// Add the node to the chain.
+	// There are a few possibilities here:
+	//  1) This node is a child of an existing block node
+	//  2) This node is the parent of one or more nodes
+	//  3) Neither 1 or 2 is true which implies it's an orphan block and
+	//     therefore is an error to insert into the chain
+	prevHash := &blockHeader.PrevBlock
+	if parentNode, ok := bi.index[*prevHash]; ok {
+		// Case 1 -- This node is a child of an existing block node.
+		// Update the node's work sum with the sum of the parent node's
+		// work sum and this node's work, append the node as a child of
+		// the parent node and set this node's parent to the parent
+		// node.
+		node.workSum = node.workSum.Add(parentNode.workSum, node.workSum)
+		parentNode.children = append(parentNode.children, node)
+		node.parent = parentNode
+
+	} else if childNodes, ok := bi.depNodes[*hash]; ok {
+		// Case 2 -- This node is the parent of one or more nodes.
+		// Update the node's work sum by subtracting this node's work
+		// from the sum of its first child, and connect the node to all
+		// of its children.
+		node.workSum.Sub(childNodes[0].workSum, node.workSum)
+		for _, childNode := range childNodes {
+			childNode.parent = node
+			node.children = append(node.children, childNode)
+		}
+
+	} else {
+		// Case 3 -- The node doesn't have a parent in the node cache
+		// and is not the parent of another node.  This means an arbitrary
+		// orphan block is trying to be loaded which is not allowed.
+		str := "loadBlockNode: attempt to insert orphan block %v"
+		return nil, AssertError(fmt.Sprintf(str, hash))
+	}
+
+	// Add the new node to the indices for faster lookups.
+	bi.index[*hash] = node
+	bi.depNodes[*prevHash] = append(bi.depNodes[*prevHash], node)
+
+	return node, nil
+}
+
+// PrevNodeFromBlock returns a block node for the block previous to the
+// passed block (the passed block's parent).  When it is already in the memory
```

Comment:
suggestion: `When it is already in the memory block chain` could be simplified to  `When it is already in the block index`

Reply:
This is in #990.  Fixed there.